### ISSUE TITLE
Convert model instance to JSON

### ISSUE THREAD

**Issue by karloscodes** [pynamodb/PynamoDB#152]
Is there a way to convert a model instance to JSON??, I'm getting `<Obj> is not JSON serializable` when trying to serialize with JSON module

**Comment by csimmons0** [comment]
Hi, Carlos. I just submitted a pull request (#156) that may provide what you're looking for. It lets you convert a model instance to a Python dictionary, and you can easily convert a dictionary to JSON.

**Comment by yedpodtrzitko** [comment]
@ccverak you can use `Model()._serialize()`

**Comment by jdno** [comment]
As @csimmons0 said in his pull request: `_serialize()` is too close to the data representation in DynamoDB and thus not suitable for (my) APIs. Besides, `_serialize()` looks like a  "private" method and should not be used by third-party code.

I agree that we need a method that produces a dict or JSON representation of the model object that abstracts from the underlying data representation. For example, here is what I get when I serialize my object with `_serialize()`:

```
{
  'attributes': {
    'owner': {'S': u'test|12345678'},
    'updated': {'S': u'2016-12-19T14:52:32.140080+00:00'},
    'name': {'S': u'Test Organization'},
    'members': {'SS': [u'test|12345678']},
    'created': {'S': u'2016-12-19T14:52:32.140106+00:00'}
  },
  'HASH': u'59fc2bcd-230d-427c-ae9f-9834bf29fc8a'
}
```

However, the following is what I need to pass it along to my user-facing API:

```python
{
  'id': '59fc2bcd-230d-427c-ae9f-9834bf29fc8a',
  'name': 'Test Organization',
  'owner': 'test|12345678',
  'members': [
    'test|12345678'
  ],
  'created': '2016-12-19T14:52:32.140106+00:00',
  'updated': '2016-12-19T14:52:32.140106+00:00'
}
```

Would be great to get some feedback on this from @jlafon so that we can merge (or work on) #156.

**Comment by mathom** [comment]
@jdno and @ccverak:

I'm using this in my project - I hope it helps if you're still stuck!

```
class ModelEncoder(json.JSONEncoder):
    def default(self, obj):
        if hasattr(obj, 'attribute_values'):
            return obj.attribute_values
        elif isinstance(obj, datetime.datetime):
            return obj.isoformat()
        return json.JSONEncoder.default(self, obj)


def json_dumps(obj):
    return json.dumps(obj, cls=ModelEncoder)
```

**Comment by kadrach** [comment]
You can also utilize the built-in serializers, see e.g. [here](https://github.com/jlafon/PynamoDB/blob/24e14caef9f0588ea7fcb89e4742fa3253a1b667/pynamodb/attributes.py#L370) for `datetime`. Something along the lines of 

```python
class Model(pynamodb.models.Model):
    ...

    def __iter__(self):
        for name, attr in self._get_attributes().items():
            yield name, attr.serialize(getattr(self, name))
```
`dict(model_instance)` should return the serialized object. Works for my simple usecase, not extensively tested  :smile:

**Comment by toshke** [comment]
by simple extension or using base class, it is easy to get dict for the dynamo item, which can be furhter serialised.. i tried @mathom  answer, but it did not worked for `MapAttribute` type. However just extending models from base class below was enough to extract dict and further serialise to JSON/xml

```
class BaseModel(Model):
    def to_dict(self):
        rval = {}
        for key in self.attribute_values:
            rval[key] = self.__getattribute__(key)
        return rval

```

**Comment by gabrielBusta** [comment]
Yet another possible solution.

Each instance of a model is serialized into a JSON list where the hashing key is the 1st element. The 2nd element is a JSON object with the value of each one of the model's attributes.

i.e.
[
    "hashing key",
    {
        "attribute 1": "value 1",
        "attribute 2": "value 2",
        ...
        "attribute n": "value n"
    }
]

This approach has been working well for me so far.

```python
import json
from pynamodb.models import Model


def head(iterable):
    return iterable[0]


class EZJSONModel(Model):
    """
    A model that can be serialized to and from JSON easily.
    """
    def serialize(self):
        """
        :return: a json list in the following format [hash_key, attribute_dict]
        """
        hash_key, data = self._get_json()
        data = data['attributes']
        data = {
            attr: head(list(data[attr].values())) for attr in data
        }
        return json.dumps([hash_key, data])
```

**Comment by chrisnorman27** [comment]
I expanded on @kadrach's solution - this worked for my table which contained some MapAttributes as well as NumberAttributes and UnicodeAttributes.

```python
class MyModel(pynamodb.models.Model)
...
    def __iter__(self):
        for name, attr in self.get_attributes().items():
            if isinstance(attr, MapAttribute):
                yield name, getattr(self, name).as_dict()
            else:
                yield name, attr.serialize(getattr(self, name))
```

`dict(model_instance)` then returned a dict of the model. Untested for models containing ListAttributes!

**Comment by codeocelot** [comment]
There's a small typo in @chrisnorman27's snippet.

```diff
-         for name, attr in self.get_attributes().items():
+         for name, attr in self._get_attributes().items():
```

From his solution, this how I added support for ListAttributes: 

```py
class MyModel(pynamo.models.Model):
    def __iter__(self):
        for name, attr in self._get_attributes().items():
            if isinstance(attr, MapAttribute):
                yield name, getattr(self, name).as_dict()
            if isinstance(attr, ListAttribute):
                yield name, [el.as_dict() for el in getattr(self, name)]
            else:
                yield name, attr.serialize(getattr(self, name))
```

I feel like this is something this library should natively support, or least document.  It's easy to get sucked up into thinking `deserialize`/`serialize` methods are for JSON serialization.

**Comment by mikegrima** [comment]
Does anyone know how to deal with nested values? In my case, I'm having an problem with a nested `Decimal` inside of a `MapAttribute` :/

**Comment by bendog** [comment]
I've added a numeric test to return numbers as numerics as this understandably was causing issues with sorting.

```python
    def __iter__(self):
        for name, attr in self.get_attributes().items():
            if isinstance(attr, attributes.MapAttribute):
                if getattr(self, name):
                    # app.logger.debug(name)
                    yield name, getattr(self, name).as_dict()
            elif isinstance(attr, attributes.UTCDateTimeAttribute):
                if getattr(self, name):
                    yield name, attr.serialize(getattr(self, name))
            elif isinstance(attr, attributes.NumberAttribute):
                # if numeric return value as is.
                yield name, getattr(self, name)
            else:
                yield name, attr.serialize(getattr(self, name))
```

**Comment by gowtham800** [comment]
I am getting empty attribute object response. Can some one help me please....

class PNFGDINDEX(GlobalSecondaryIndex):
                """
                This class represents a global secondary index
                """
                class Meta:
                    # index_name is optional, but can be provided to override the default name
                    index_name = 'PN-FGD-Index'
                    read_capacity_units = 5
                    write_capacity_units = 5
                    # All attributes are projected
                    #projection = IncludeProjection('Id, createdTimeStamp, ProjectName, FileGeneratedDate, FileNameKey, VehicleName')
                    projection = AllProjection()

                # This attribute is the hash key for the index
                # Note that this attribute must also exist
                # in the model
                ProjectName = UnicodeAttribute(hash_key=True)
                FileGeneratedDate = UnicodeAttribute(range_key=True)

            class TestModel(Model):
                """
                A test model that uses a global secondary index
                """
                class Meta:
                    aws_access_key_id = "xxxxxxxxx"
                    aws_secret_access_key = "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
                    table_name = 'TEST_EDCBUILDDATA1'
                    region = 'ap-south-1'
                    
                Id = UnicodeAttribute(hash_key=True)
                createdTimeStamp = UnicodeAttribute(range_key=True)
                ProjectName = UnicodeAttribute
                FileGeneratedDate = UnicodeAttribute
                FileNameKey = UnicodeAttribute
                VehicleName = UnicodeAttribute
                view_index = PNFGDINDEX()
                

            for item in TestModel.view_index.query(pn):
                #print("Item queried from index: {0}".format(item))
                print(item._serialize())
                FileNameParquetList.append(item.FileNameKey)

**Comment by shungok** [comment]
Here is my solution for multi tier structure.

```python
from datetime import datetime
import json

from pynamodb.models import Model
from pynamodb.attributes import MapAttribute

class BaseModel(Model):
    def to_json(self, indent=2):
        return json.dumps(self.to_dict(), indent=indent)
        
    def to_dict(self):
        ret_dict = {}
        for name, attr in self.attribute_values.items():
            ret_dict[name] = self._attr2dic(attr)
        
        return ret_dict

    def _attr2dic(self, attr):
        # compare with list class. It is not ListAttribute.
        if isinstance(attr, list):
            _list = []
            for l in attr:
                _list.append(self._attr2dic(l))
            return _list
        elif isinstance(attr, MapAttribute):
            _dict = {}
            for k, v in attr.attribute_values.items():
                _dict[k] = self._attr2dic(v)
            return _dict
        elif isinstance(attr, datetime):
            return attr.isoformat()
        else:
            return attr
```

**Comment by Swington** [comment]
Because my models had many nested attributes (e.g. my model had a list of map, which contained a map) none of the solutions described above seemed to work. 

I was able to compose a solution using `dynamodb-json` module, though:
```python
json_util.loads(model._serialize()['attributes'])
```
The only inconvenience here is that Keys are kept separately from attributes, you have to compose a common dict from all of those. As an instance method, of a class with single Hash Key, it would like something like that:
```python
class BaseModel(Model):
    def to_dict(self):
        serialized_self = self._serialize()
        serialized_self['attributes'].update({'clientId': serialized_self['HASH']})
        return json_util.loads(serialized_self['attributes'])
```

**Comment by hassan-bazzi** [comment]
If you have a custom MapAttribute you can use .as_dict() i.e.

`YourModel(user_id).application.as_dict()`

Where `application` is a custom attribute that extends MapAttribute.

The only issue with that was that any ListAttributes embedded in another MapAttribute inside of our `application` MapAttribute don't jsonify properly. I made a custom BaseMapAttribute class and used the solution here: 

https://github.com/pynamodb/PynamoDB/issues/491

To override as_dict and make it transform the objects in the list as dicts

**Comment by jpinner-lyft** [comment]
Closed by #857
